docs/coding-conventions.md

```diff
@@ -0,0 +1,297 @@
+
+## Write elegant, cohesive and easily readable code
+
+- If reviewers ask questions about why the code is the way it is, that's a
+  sign that your code is not clear enough.
+- Try improving your code on the expense of writing comments. A properly-named
+  function is better than a comment on a code stanza.
+- Add comments where code-documentation is not enough.
+
+## Avoid nesting and complexity by using early returns
+
+- Deeply nested if/else statements make it harder to understand code.
+- By using early returns you can avoid nesting of if/else statements.
+- Code that is easier to understand will be easier to review and less
+  likely to contain hidden bugs (e.g. control flow issues).
+
+### Bad example
+
+```go
+val, err := doSomething()
+if err == nil {
+    if val {
+        return doSomethingElse()
+    } else {
+        return doAnotherThing()
+    }
+} else {
+    return err
+}
+```
+
+### Good example
+
+```go
+val, err := doSomething()
+if err != nil {
+    return err
+}
+
+if val {
+    return doSomethingElse()
+}
+
+return doAnotherThing()
+```
+
+## Use the same coding style throughout a file
+
+- When adding to a file, stick to the existing coding style to make reviews
+  easier.
+- When you encounter code that requires a larger cleanup do it in a separate
+  commit or PR.
+
+### Bad example
+
+```go
+func checkAB(in string) bool {
+    var bool ok
+    if in == "A" || in == "B" {
+        ok = true
+    } else if in == "C" {
+        ok = false
+    }
+    return ok
+}
+
+func checkBC(in string) bool {
+    switch in {
+    case "B", "C":
+        return true
+    }
+
+    return false
+}
+```
+
+### Good example
+
+```go
+func checkAB(in string) bool {
+    return in == "A" || in == "B"
+}
+
+func checkBC(in string) bool {
+    return in == "B" || in == "C"
+}
+```
+
+# Avoid repetition of hardcoded values
+
+- Use constants if you would repeat hardcoded values otherwise.
+- Keep them private if they are used only in a single package.
+- Carefully consider if you should make a constant exported.
+
+### Bad example
+
+```go
+func getImportantAnnotation(obj metav1.ObjectMeta) string {
+    return obj.GetAnnotations()["kubevirt.io/my-annotation"]
+}
+
+func setImportantAnnotation(obj metav1.ObjectMeta, val string) {
+    obj.GetAnnotations()["kubevirt.io/my-annotation"] = val
+}
+```
+
+### Good example
+
+```go
+const annotationKey = "kubevirt.io/my-annotation"
+
+func getImportantAnnotation(obj metav1.ObjectMeta) string {
+    return obj.GetAnnotations()[annotationKey]
+}
+
+func setImportantAnnotation(obj metav1.ObjectMeta, val string) {
+    obj.GetAnnotations()[annotationKey] = val
+}
+```
+
+# Uniform import order and naming
+
+- Use the following import order with one block per item:
+    - Golang standard libraries
+    - Ginkgo / Gomega imports (only in test files)
+    - Third-party libraries
+    - Local packages
+- Use the following naming schemes for imports:
+    - `virtv1` for imports of `kubevirt.io/api`
```

Comment:
I've seen uses of both and IMO I prefer `virtv1. What would you suggest to mention here instead? Just the uniform import order? Would you change this item to be more about formatting in general?

Reply:
My general rule is that if I write the code in the kubevirt codebase, I use the short `v1`, because I do not need to express it is "my" API.
When I use 3rd party ones, like k8s ones, I better mention from where it is.

If you do not feel this is a good general guide, then perhaps there is no right answer and we can just not recommend anything about aliasing.

If there is some guide somewhere (k8s?) that has an establish rule, then we could embrace it.

Reply:
I did not find a guide in the K8S docs, but it makes sense to use the shorter names for local code. Updated it.